litestar/openapi/plugins.py

```diff
@@ -0,0 +1,552 @@
+        **kwargs: Any,
+    ) -> None:
+        """Initialize the OpenAPI UI render plugin.
+
+        Args:
+            path: Path to serve the OpenAPI UI at.
+            media_type: Media type for the handler.
+            **kwargs: Additional arguments to pass to the base class.
+        """
+        super().__init__(path=path, media_type=media_type, **kwargs)
+
+    def render(self, request: Request, openapi_schema: dict[str, Any]) -> bytes:
+        """Render an OpenAPI schema as YAML.
+
+        Args:
+            request: The request.
+            openapi_schema: The OpenAPI schema as a dictionary.
+
+        Returns:
+            The rendered OpenAPI schema as YAML.
+        """
+        # using msgspec.to_builtins() ensures that any examples generated by polyfactory that have the
+        # UNSET value (possible if the examples are being generated for a partial DTO model which makes
+        # every type a union with UNSET) are stripped out.
+        openapi_schema = msgspec.to_builtins(
+            openapi_schema, enc_hook=get_serializer(request.route_handler.resolve_type_encoders())
+        )
+        return yaml.dump(openapi_schema, default_flow_style=False).encode("utf-8")
+
+
+class RapidocRenderPlugin(OpenAPIRenderPlugin):
+    def __init__(
+        self,
+        *,
+        version: str = "9.3.4",
+        js_url: str | None = None,
+        path: str | tuple[str, ...] = "/rapidoc",
+        **kwargs: Any,
+    ) -> None:
+        """Initialize the OpenAPI UI render plugin.
+
+        Args:
+            version: Rapidoc version to download from the CDN. If js_url is provided, this is ignored.
+            js_url: Download url for the RapiDoc JS bundle. If not provided, the version will be used to construct the
+                url.
+            path: Path to serve the OpenAPI UI at.
+            **kwargs: Additional arguments to pass to the base class.
+        """
+        self.js_url = js_url or f"https://unpkg.com/rapidoc@{version}/dist/rapidoc-min.js"
+        super().__init__(path=path, **kwargs)
+
+    def render(self, request: Request, openapi_schema: dict[str, Any]) -> bytes:
+        """Render an HTML page for Rapidoc.
+
+        Notes:
+            - override this method to customize the template.
+
+        Args:
+            request: The request.
+            openapi_schema: The OpenAPI schema as a dictionary.
+
+        Returns:
+            A rendered html string.
+        """
+
+        head = f"""
+          <head>
+            <title>{openapi_schema["info"]["title"]}</title>
+            {self.favicon}
+            <meta charset="utf-8"/>
+            <meta name="viewport" content="width=device-width, initial-scale=1">
+            <script src="{self.js_url}" crossorigin></script>
+            <style>{self.style}</style>
+          </head>
+        """
+
+        body = """
+          <body>
+            <rapi-doc spec-url="openapi.json" />
```

Comment:
When we render the UI with a callback to retrieve the JSON, we require that the `openapi.json` endpoint is set in `OpenAPIConfig.enabled_endpoints`, but there is no actual checking or control of that dependency.

This is not new with this plugin-based implementation, but prob worth thinking about all the same.

Reply:
> I think we should just always include the `JsonRenderPlugin` so that `openapi.json` is always available. It is a pretty efficient endpoint, and doesn't require any extra dependencies.

Yeah, JSON should be enabled by default. Everything else is optional.